ArmPkg/Drivers/SmmuDxe/IoMmu.c

```diff
@@ -0,0 +1,523 @@
+#define PAGE_TABLE_READ_BIT         (0x1 << 6)
+#define PAGE_TABLE_WRITE_BIT        (0x1 << 7)
+#define PAGE_TABLE_ENTRY_VALID_BIT  0x1
+#define PAGE_TABLE_BLOCK_OFFSET     0xFFF
+#define PAGE_TABLE_ACCESS_FLAG      (0x1 << 10)
+#define PAGE_TABLE_DESCRIPTOR       (0x1 << 1)
+#define PAGE_TABLE_INDEX(VA, LEVEL)  (((VA) >> (12 + (9 * (PAGE_TABLE_DEPTH - 1 - (LEVEL))))) & 0x1FF)
+
+EDKII_IOMMU_PROTOCOL  SmmuIoMmu = {
+  EDKII_IOMMU_PROTOCOL_REVISION,
+  IoMmuSetAttribute,
+  IoMmuMap,
+  IoMmuUnmap,
+  IoMmuAllocateBuffer,
+  IoMmuFreeBuffer,
+};
+
+typedef struct IOMMU_MAP_INFO {
+  UINTN     NumberOfBytes;
+  UINT64    VA;
+  UINT64    PA;
+} IOMMU_MAP_INFO;
+
+/**
+  Update the flags of a page table entry.
+
+  @param [in]  Table         Pointer to the page table.
+  @param [in]  SetFlagsOnly  Boolean to indicate if only flags should be set.
+  @param [in]  Flags         Flags to set or clear.
+  @param [in]  Index         Index of the entry to update.
+
+  @retval EFI_SUCCESS        Success.
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+UpdateFlags (
+  IN PAGE_TABLE  *Table,
+  IN BOOLEAN     SetFlagsOnly,
+  IN UINT64      Flags,
+  IN UINT64      Index
+  )
+{
+  if (Table == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  // This boolean is used to explicity update the R/W bits in the page table entry.
+  // Allows clearing the R/W bits without affecting the other bits in the entry.
+  if (SetFlagsOnly) {
+    if (Flags != 0) {
+      // Set R/W bits in page table entry
+      Table->Entries[Index] |= Flags;
+    } else {
+      // Clear R/W bits in page table entry
+      Table->Entries[Index] &= ~(PAGE_TABLE_READ_BIT | PAGE_TABLE_WRITE_BIT);
+    }
+  } else {
+    // Set R/W bits in page table entry
+    Table->Entries[Index] |= Flags;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Update the mapping of a virtual address to a physical address in the page table.
+
+  Iterates through the page table levels to find the leaf entry for the given virtual address and
+  validates entries along the way as needed. The leaf entry is then updated with the physical address along
+  with appropriate flags and valid bit set. The option SetFlagsOnly allows traversal of the page table while
+  only updating the flags of the entry, allowing clearing of flag bits as well.
+
+  @param [in]  Root          Pointer to the root page table.
+  @param [in]  VA            Virtual address to map.
+  @param [in]  PA            Physical address to map to.
+  @param [in]  Flags         Flags to set for the mapping.
+  @param [in]  Valid         Boolean to indicate if the entry is valid.
+  @param [in]  SetFlagsOnly  Boolean to indicate if only flags should be set.
+
+  @retval EFI_SUCCESS        Success.
+  @retval EFI_INVALID_PARAMETER  Invalid parameter.
+  @retval EFI_OUT_OF_RESOURCES   Out of resources.
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+UpdateMapping (
+  IN PAGE_TABLE  *Root,
+  IN UINT64      VA,
+  IN UINT64      PA,

```

Comment:
The style convention is to avoid abbreviations with Pascal case.

```suggestion
  IN UINT64      VirtualAddress,
  IN UINT64      PhysicalAddress,
```

Reply:
Fixed